### PR TITLE
Userclass

### PR DESCRIPTION
**Pull request by jjalfaro9**
So I changed the name of the mongoDB database from MoviesDatabase to MovieMatcherDatabase since we are also storing User data. 

We have decided to work with BCrypt in order to store an encrypted version of  users' passwords.

User class has been, for the most part, fully implemented. Need to work on methods that will use a Group class. Also, need to finish resetPassword and login methods. 

Last thing, need to work on saving user ratings!

### CODE REVIEW

**Comment by jjalfaro9** [comment]
Another note: 
The friends data member is an ArrayList<User> and the groups data member is an ArrayList<Group>. So for the time being I just added these data members into mongoDB. I don't think this will work. I need to write some tests. I just wan't sure how mongoDB stores lists.

File: src/main/java/com/johnstarich/moviematcher/models/User.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    public User load() {
+        Document user = getCollection().find(eq("_id", _id)).first();
+        Gson gson = new GsonBuilder().create();
+        return gson.fromJson(user.toJson(), User.class);
+    }
+
+    public User register(String password) throws HttpException {
+        if(this.load().equals(this)) { //need to test this, too see if it actually works!! ...
```
Comment (JohnStarich) [inline src/main/java/com/johnstarich/moviematcher/models/User.java]: Double check this works before we merge 😄 
Make sure that if it throws an Exception we catch it and throw the better HttpException instead.

File: src/main/java/com/johnstarich/moviematcher/models/User.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+                        .append("groups", true)
+                        .append("password", true)
+                        .append("score", new Document("$meta", "textScore"))
+                ),
+                new Document("$sort", new Document("score",-1))
+            )
+        );
+
+        ArrayList<User> queryResults = new ArrayList<User>();
+        Gson gson = new GsonBuilder().create();
```
Comment (JohnStarich) [inline src/main/java/com/johnstarich/moviematcher/models/User.java]: Make this a static instance inside the class so it isn't created on every method call

File: src/main/java/com/johnstarich/moviematcher/models/User.java [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        iterable
+                .map(document -> gson.fromJson(document.toJson(), User.class))
+                .forEach((Block<User>) u -> queryResults.add(u));
+
+        return queryResults;
+    }
+
+    public static User loadByUsername(String email) {
+        Document user = getCollection().find(eq("email", email)).first();
+        Gson gson = new GsonBuilder().create();
```
Comment (JohnStarich) [inline src/main/java/com/johnstarich/moviematcher/models/User.java]: reuse aforementioned gson instance

File: src/main/java/com/johnstarich/moviematcher/models/User.java [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    public boolean equals(Object o) { return o == this || o instanceof User && ((User) o)._id == _id && ((User) o).email == email; }
+
+    @Override
+    public int hashCode() {
+        return _id.hashCode();
+    }
+
+    public User load() {
+        Document user = getCollection().find(eq("_id", _id)).first();
+        Gson gson = new GsonBuilder().create();
```
Comment (JohnStarich) [inline src/main/java/com/johnstarich/moviematcher/models/User.java]: reuse mentioned gson instance

**Comment by JohnStarich** [comment]
LGTM :+1: please review my additions (there should be a "changes since you've viewed this" button for your convenience)
lmk when you think my changes are good!

**Comment by jjalfaro9** [comment]
LGTM. :+1:  :)

**Pull request merged**